sockeye/utils.py

```diff
@@ -271,49 +272,99 @@ def get_num_gpus() -> int:
 
 
 @contextmanager
-def acquire_gpu(lock_dir: str = "/var/lock", retry_wait: int = 10):
+def acquire_gpus(num_gpus:int, lock_dir: str = "/tmp", retry_wait_min: int = 30, retry_wait_rand: int = 30):
     """
-    Acquires a GPU by locking a file (therefore this assumes that everyone using GPUs calls this method and shares the
-    lock directory).
+    Acquire a number of GPUs in a transactional way. Will try to acquire all the requested number of GPUs. If currently
+    not enough GPUs are available all locks will be released and we wait until we retry. Will retry until enough
+    GPUs become available.
+
+    :param num_gpus: The number of GPUs to acquire.
+    :param lock_dir: The directory for storing the lock file.
+    :param retry_wait_min: The minimum number of seconds to wait between retries.
+    :param retry_wait_rand: Randomly add between 0 and `retry_wait_rand` seconds to the wait time.
+    :return: yields a list of GPU ids.
+    """
+    num_gpus_available = get_num_gpus()
+
+    if num_gpus > num_gpus_available:
+        raise ValueError("Requested %d GPUs, but only %d are available." % (num_gpus, num_gpus_available))
+
+    logger.info("Attempting to acquire %d GPUs of %d GPUs.", num_gpus, num_gpus_available)
+
+    while True:
+        with ExitStack() as exit_stack:
+            acquired_gpus = []
+            for _ in range(num_gpus):
+                gpu_id = exit_stack.enter_context(acquire_gpu(lock_dir))
+                if gpu_id is not None:
+                    acquired_gpus.append(gpu_id)
+                else:
+                    break
+            if len(acquired_gpus) == num_gpus:
+                try:
+                    yield acquired_gpus
```

Comment:
if we want to use it in a with clause in train.py to automatically release them we still need to the yield. Fun fact: because the previous version of the code was not wrapped in a try: finally block the locks were not actually ever release. This wasn't a huge issue as the operating systems file system anyway releases them when the process dies, but I guess it's not very clean.